libyul/backends/evm/Dominator.h

```diff
@@ -0,0 +1,309 @@
+				if (semi[dfIdx] == std::numeric_limits<size_t>::max())
+				{
+					parent[dfIdx] = m_vertexIndices[_v];
+					_dfs(w, _dfs);
+				}
+				predecessors[m_vertexIndices[w]].insert(m_vertexIndices[_v]);
+			});
+		};
+		dfs(_entry, dfs);
+
+		// Process the vertices in decreasing order of the DFS number
+		for (size_t wIdx: m_vertices | ranges::views::reverse | ranges::views::transform(toIdx))
+		{
+			// step 3
+			// NOTE: this is an optimization, i.e. performing the step 3 before step 2.
+			// The goal is to process the bucket in the beginning of the loop for the vertex ``w``
+			// instead of ``parent[w]`` in the end of the loop as described in the original paper.
+			// Inverting those steps ensures that a bucket is only processed once and
+			// it does not need to be erased.
+			// The optimization proposal is available here: https://jgaa.info/accepted/2006/GeorgiadisTarjanWerneck2006.10.1.pdf pg.77
+			for (size_t vIdx: bucket[wIdx])
+			{
+				size_t uIdx = eval(vIdx);
+				solAssert(uIdx <= vIdx);
+				idom[vIdx] = (semi[uIdx] < semi[vIdx]) ? uIdx : wIdx;
+			}
+
+			// step 2
+			for (size_t vIdx: predecessors[wIdx])
+			{
+				size_t uIdx = eval(vIdx);
+				solAssert(uIdx <= vIdx);
+				if (semi[uIdx] < semi[wIdx])
+					semi[wIdx] = semi[uIdx];
+			}
+			bucket[semi[wIdx]].emplace_back(wIdx);
+			link(parent[wIdx], wIdx);
+			solAssert(ancestor[wIdx] == parent[wIdx]);
+		}
+
+		// step 4
+		idom[0] = 0;
+		for (size_t wIdx: m_vertices | ranges::views::drop(1) | ranges::views::transform(toIdx))
+			if (idom[wIdx] != semi[wIdx])
+				idom[wIdx] = idom[idom[wIdx]];
+
+		return idom;
+	}
+
+private:
+	/// Build dominator tree from the immediate dominators set.
+	/// The function groups all the indices that are immediately dominated by a vertex.
+	void buildDominatorTree()
+	{
+		solAssert(m_immediateDominators.size() == m_vertices.size());
+
+		// Ignoring the entry node since no one dominates it.
+		for (size_t index = 1; index < m_vertices.size(); ++index)
+		{
+			solAssert(m_immediateDominators[index] < index);
+			m_dominatorTree[m_immediateDominators[index]].emplace_back(index);
+		}
+	}
+
+	/// Keep the list of vertices in the DFS order.
+	/// i.e. m_vertices[i] is the vertex whose DFS index is i.
+	std::vector<V> m_vertices;
+
+	/// Maps Vertex to its DFS index.
+	std::map<V, size_t> m_vertexIndices;
+
+	/// Immediate dominators by index.
+	/// Maps a vertex based on its DFS index (i.e. array index) to its immediate dominator DFS index.
+	/// The entry vertex is the first element of the vector.
+	///
+	/// e.g. to get the immediate dominator of a Vertex w:
+	/// idomIdx = m_immediateDominators[m_vertexIndices[w]]
+	/// idomVertex = m_vertices[domIdx]
+	std::vector<size_t> m_immediateDominators;
+
+	/// Maps a vertex to all vertices that it dominates.
+	/// If the vertex does not dominate any other vertex it has no entry in the map.
+	std::map<size_t, std::vector<size_t>> m_dominatorTree;
```

Comment:
Why not a `set`? The order isn't meaningful here, is it?
```suggestion
	std::map<size_t, std::set<size_t>> m_dominatorTree;
```

Reply:
Since the vertices are maintained in DFS order within `m_vertices`, having a vector of vertices that are dominated in `m_dominatorTree` preserves the order of all vertices dominated by that vertex. However, I'm not sure how useful this information will be. Currently it is not used anyway. I'm fine in changing it if you see a benefit to having it as a set.

Reply:
I'm using a vertex of sets now ;)